### PR TITLE
Check indexed and included columns before signature calculation

### PR DESCRIPTION
**Pull request by sezruby**
### What is the context for this pull request?

 - **Tracking Issue**: #160 
 - **Parent Issue**: N/A
 - **Dependencies**: N/A

### What changes were proposed in this pull request?

This PR changes the order of index validation.
Currently, we gets the candidate index in the following order:

1. getCandidateIndex => signature calculation & comparison
2. index schema check => all required columns & all indexed columns
3. other conditions (e.g. compatible check in Join Index)
4. rank algorithm

However, signature calculation might take longer in case there are numerus source files. In addition, if there are a lot of indexes for different locations, file-level check without any pruning might be very inefficient.

With this PR, the index schema is firstly validated before the signature check.
1. index schema check => each candidate index should include all required columns & all indexed columns
2. getCandidateIndex => signature calculation & comparison
3. other conditions (e.g. compatible left & right indexes for Join Index)
4. rank algorithm

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Tested by FilterIndexRuleTest, JoinIndexRuleTest and E2EHyperspaceRulesTests.

### CODE REVIEW

File: src/main/scala/com/microsoft/hyperspace/index/rules/FilterIndexRule.scala [hunk 1]
```diff
@@ -152,9 +153,13 @@ object FilterIndexRule
...
           .indexCollectionManager
-        val candidateIndexes = RuleUtils.getCandidateIndexes(indexManager, r)
 
-        candidateIndexes.filter { index =>
+        // TODO: the following check only considers indexes in ACTIVE state for usage. Update
+        //  the code to support indexes in transitioning states as well.
+        //  See https://github.com/microsoft/hyperspace/issues/65
+        val allIndexes = indexManager.getIndexes(Seq(Constants.States.ACTIVE))
+
+        val candidateIndexes = allIndexes.filter { index =>
```
Comment (pirz) [inline src/main/scala/com/microsoft/hyperspace/index/rules/FilterIndexRule.scala:162]: nit: Can you add a brief comment here as why the order of code blocks matter here (first index schema check then signature computation/comparison - similar to what you already explained in the PR description)? This is to make sure nobody in a future PR moves the code blocks around and causes negative perf impact.

**Review by pirz — COMMENTED** [review]
LGTM - I added some minor points for adding comments.

File: src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala [hunk 2]
```diff
@@ -358,9 +328,13 @@ object JoinIndexRule
...
     require(resolve(spark, rRequiredIndexedCols, rRequiredAllCols).isDefined)
 
-    val lUsable = getUsableIndexes(lIndexes, lRequiredIndexedCols, lRequiredAllCols)
-    val rUsable = getUsableIndexes(rIndexes, rRequiredIndexedCols, rRequiredAllCols)
-    val compatibleIndexPairs = getCompatibleIndexPairs(lUsable, rUsable, lRMap)
+    val lUsable = getUsableIndexes(allIndexes, lRequiredIndexedCols, lRequiredAllCols)
+    val rUsable = getUsableIndexes(allIndexes, rRequiredIndexedCols, rRequiredAllCols)
+
+    val lIndexes = RuleUtils.getCandidateIndexes(lUsable, RuleUtils.getLogicalRelation(left).get)
+    val rIndexes = RuleUtils.getCandidateIndexes(rUsable, RuleUtils.getLogicalRelation(right).get)
```
Comment (pirz) [inline src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala:335]: ditto: (Similar to comment on FilterIndex rule) Please add a brief comment on why schema check should be done before signature computation/comparison.

File: src/main/scala/com/microsoft/hyperspace/index/rules/FilterIndexRule.scala (hunk 1 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/FilterIndexRule.scala:112]: Thanks added some comment below. :)

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 3]
```diff
@@ -37,13 +36,13 @@ object RuleUtils {
   /**
    * Get active indexes for the given logical plan by matching signatures.
```
Comment (apoorvedave1) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:37]: could you please update the description

Comment (apoorvedave1) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:37]: also the pre-requisite of ACTIVE indexes

**Comment by imback82** [comment]
@sezruby Can you resolve the conflict?

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 3 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala]: Thanks! I updated the comment. BTW seems active indexes are not pre-requisite for this function as we check the status with  `index => index.created` (`def created`: status == "ACTIVE")

**Comment by sezruby** [comment]
Actually there will be inevitable conflict with #171, around `getCandidateIndexes` 
I'll rebase it as soon as one of them is merged.

File: src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala [hunk 4]
```diff
@@ -341,9 +307,22 @@ object JoinIndexRule
...
-    val compatibleIndexPairs = getCompatibleIndexPairs(lUsable, rUsable, lRMap)
+    val lUsable = getUsableIndexes(allIndexes, lRequiredIndexedCols, lRequiredAllCols)
+    val rUsable = getUsableIndexes(allIndexes, rRequiredIndexedCols, rRequiredAllCols)
+
+    // Get candidate via file-level metadata validation. This is performed after pruning
+    // by column schema, as this might be expensive when there are numerous files in the
+    // relation or many indexes to be checked.
+    val lIndexes = RuleUtils.getCandidateIndexes(
+      lUsable,
+      RuleUtils.getLogicalRelation(left).get,
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala:318]: is it guaranteed that `getLogicalRelation` would always return `Some` at this point?

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala:320]: Thanks! I added the condition in `isApplicable`

**Review by imback82 — APPROVED** [review]
LGTM, thanks @sezruby!

**Pull request merged**